### PR TITLE
[jax2tf] Add support for first-order AD to converted functions

### PR DESCRIPTION
**Pull request by gnecula**
On conversion, optionally, we convert the jax.vjp of the converted
function and set the result as tf.custom_gradient.

Only first-order diff is supported for now.

### CODE REVIEW

File: jax/experimental/jax2tf/jax2tf.py [hunk 1]
```diff
@@ -103,67 +103,79 @@ def add_unit(v: TfValOrUnit, aval: core.AbstractValue):
...
-      return out, grad_disabled
-
-    return inner_wrapper(*tensor_args)
-
-  return wrapper
-
-
-def convert(fun):
+def convert(fun,
+            with_gradient=False):
```
Comment (tomhennigan) [inline jax/experimental/jax2tf/jax2tf.py:107]: I feel like `True` might be a safer default to start with? Also nit that we don't need to split the line here.

File: jax/experimental/jax2tf/jax2tf.py [hunk 2]
```diff
@@ -103,67 +103,79 @@ def add_unit(v: TfValOrUnit, aval: core.AbstractValue):
...
+        raise ValueError("Unexpected variables used in forward pass. "
+                         "This should not happen for first-order differentiation. "
+                         f"variables={variables}")
+
+      def fun_vjp_jax(args_jax, out_cts_jax):
+        # One may think that we can get the pullback while we are converting
+        # the main function in the first place. That is problematic, because the
+        # pullback may contain captured tracers from the conversion of the
+        # main function. Those tracers will confuse the conversion of the
+        # pullback. So, we construct the vjp anew.
```
Comment (tomhennigan) [inline jax/experimental/jax2tf/jax2tf.py:154]: I'm not sure I fully understand the implications of this, but does this mean if I have a custom gradient that closes over some expensive to compute intermediate value from forward and uses it as part of computing the gradient that when running via jax2tf we will actually always recompute this value (because the custom gradient for our function is "disconnected" from the forward pass)?

As a concrete example for this pseudocode:

```
@custom_gradient
def op(x):
  intermediate = x ** 2
  def grad(g):
    return 5 * g * intermediate
  return intermediate ** 2
```

I think we would produce a tf graph with two copies of the forward pass, one to compute the forward result and one in order for us to compute grad (aka. we'll recompute `intermediate` in order to compute `grad`). A smart enough compiler could fix this for us I guess. I'm a bit worried because this is a pretty common use case for custom gradients so we should at least call it out as a major limitation.

**Review by tomhennigan — APPROVED** [review]
This is a big improvement! One question below, but I think this can go in regardless.

File: jax/experimental/jax2tf/jax2tf.py (hunk 2 above)
Comment (gnecula) [reply jax/experimental/jax2tf/jax2tf.py:154]: You are right about this performance issue. It is in fact a bit worse, in the sense that you pay this cost even in the absence of custom gradients. This is likely fixable, but will require somewhat deeper messing with tracers and linear_util. I plan to look at it later, for now I wanted to get the main pieces in place.

File: jax/experimental/jax2tf/jax2tf.py (hunk 1 above)
Comment (gnecula) [reply jax/experimental/jax2tf/jax2tf.py:107]: Will fix the newline. 

I did think about the default, and I see your point. I thought of disabling gradients by default because it leads to somewhat better behaviour once you use a savedmodel. Currently a restored saved model that contains custom gradients fails in obscure ways. If we disable the gradients, we add a PreventGradient op, which is saved in saved model and gives a nice error message. 

Once we fix the handling of gradients in savedmodel, even if by inserting a PreventGradient in lieu of a custom gradient, then I think we can change the default. That is a safe change, because people who did not pass the `with_gradient` will then find themselves with code that supports AD but their code certainly does not rely on gradients or else they would have had errors.
